File: AES-CTR.py
from Crypto.Cipher import AES
from Crypto.Random import get_random_bytes
from Crypto.Util import Counter
import binascii
import time
import os

# Khóa bí mật, bạn cần lưu ý rằng khóa này phải được bảo mật cẩn thận
secret_key = get_random_bytes(16)
def int_of_string(s):
    return int(binascii.hexlify(s), 16)

def encrypt(plain_text, key):
    iv = os.urandom(16)
    ctr = Counter.new(128, initial_value=int_of_string(iv))
    aes = AES.new(key, AES.MODE_CTR, counter=ctr)
    return iv + aes.encrypt(plain_text)

def decrypt(cipher_text, key):
    iv = cipher_text[:16]
    ctr = Counter.new(128, initial_value=int_of_string(iv))
    aes = AES.new(key, AES.MODE_CTR, counter=ctr)
    return aes.decrypt(cipher_text[16:])
# Ví dụ mã hóa và giải mã
plain_text = "This is a secret message"
plain_text = plain_text.encode('utf-8')

start_time = time.time()
cipher_text = encrypt(plain_text, secret_key)
end_time = time.time()
print("Encryption Time:", (end_time - start_time)*1000 , "ms")

# Do thoi gian danh cho giai ma
start_time = time.time()
decrypted_text = decrypt(cipher_text, secret_key)
end_time = time.time()
print("Decryption Time:", (end_time - start_time)*1000 , "ms")

# Chuyển dữ liệu giải mã thành chuỗi
decrypted_text = decrypted_text.decode('utf-8')

print("Ciphertext:", cipher_text)
print("Decrypted Text:", decrypted_text)

File: AES-CTR.js
const crypto = require('crypto');
const { performance } = require('perf_hooks');

// Tạo một key và IV (vector khởi đầu)
const key = crypto.randomBytes(16); // Kích thước key là 16 bytes
const iv = crypto.randomBytes(16);  // Kích thước IV cũng là 16 bytes

// Mã hóa thông điệp
function encrypt(text, key, iv) {
  const cipher = crypto.createCipheriv('aes-128-ctr', key, iv);
  let encrypted = cipher.update(text, 'utf8', 'hex');
  encrypted += cipher.final('hex');
  return encrypted;
}

// Giải mã thông điệp
function decrypt(encrypted, key, iv) {
  const decipher = crypto.createDecipheriv('aes-128-ctr', key, iv);
  let decrypted = decipher.update(encrypted, 'hex', 'utf8');
  decrypted += decipher.final('utf8');
  return decrypted;
}

const plaintext = 'This is a secret message';

// Đo thời gian mã hóa
const startTimeEncrypt = performance.now();
const encryptedText = encrypt(plaintext, key, iv);
const endTimeEncrypt = performance.now();

console.log('Encryption Time:', endTimeEncrypt - startTimeEncrypt, 'ms');

// Đo thời gian giải mã
const startTimeDecrypt = performance.now();
const decryptedText = decrypt(encryptedText, key, iv);
const endTimeDecrypt = performance.now();

console.log('Decryption Time:', endTimeDecrypt - startTimeDecrypt, 'ms');

console.log('Encrypted Text:', encryptedText);
console.log('Decrypted Text:', decryptedText);
